File: packages/js/browser/src/robots.js
/**
 * robots.txt parser. Heuristic allow/deny only — never a crawl-budget or legal opinion.
 */

export function parseRobotsTxt(text, userAgent = "*") {
  const groups = [];
  let current = { agents: [], allow: [], disallow: [] };
  const flush = () => {
    if (current.agents.length) groups.push(current);
    current = { agents: [], allow: [], disallow: [] };
  };
  for (const raw of String(text ?? "").split(/\r?\n/)) {
    const line = raw.replace(/#.*$/, "").trim();
    if (!line) continue;
    const idx = line.indexOf(":");
    if (idx < 1) continue;
    const key = line.slice(0, idx).trim().toLowerCase();
    const value = line.slice(idx + 1).trim();
    if (key === "user-agent") {
      if (current.allow.length || current.disallow.length) flush();
      current.agents.push(value.toLowerCase());
    } else if (key === "allow") current.allow.push(value);
    else if (key === "disallow") current.disallow.push(value);
  }
  flush();
  const ua = String(userAgent || "*").toLowerCase();
  const matched = groups.filter((group) =>
    group.agents.some((agent) => agent === "*" || ua.includes(agent) || agent.includes(ua)),
  );
  return matched.length ? matched : groups.filter((group) => group.agents.includes("*"));
}

export function isRobotsAllowed(text, url, userAgent = "*") {
  let path = "/";
  try {
    path = new URL(url).pathname || "/";
  } catch {
    path = String(url || "/");
  }
  const groups = parseRobotsTxt(text, userAgent);
  if (!groups.length) return true;
  let decision = true;
  let best = -1;
  for (const group of groups) {
    for (const rule of group.disallow) {
      if (rule === "") continue;
      if (path.startsWith(rule) && rule.length >= best) {
        decision = false;
        best = rule.length;
      }
    }
    for (const rule of group.allow) {
      if (rule && path.startsWith(rule) && rule.length >= best) {
        decision = true;
        best = rule.length;
      }
    }
  }
  return decision;
}

File: python/packages/handoffkit/handoffkit/browser/robots.py
"""robots.txt parser. Heuristic allow/deny only — never a crawl-budget or legal opinion."""

from __future__ import annotations

from urllib.parse import urlparse


def parse_robots_txt(text: str, user_agent: str = "*") -> list[dict[str, list[str]]]:
    groups: list[dict[str, list[str]]] = []
    current: dict[str, list[str]] = {"agents": [], "allow": [], "disallow": []}

    def flush() -> None:
        nonlocal current
        if current["agents"]:
            groups.append(current)
        current = {"agents": [], "allow": [], "disallow": []}

    for raw in (text or "").splitlines():
        line = raw.split("#", 1)[0].strip()
        if not line or ":" not in line:
            continue
        key, value = line.split(":", 1)
        key = key.strip().lower()
        value = value.strip()
        if key == "user-agent":
            if current["allow"] or current["disallow"]:
                flush()
            current["agents"].append(value.lower())
        elif key == "allow":
            current["allow"].append(value)
        elif key == "disallow":
            current["disallow"].append(value)
    flush()
    ua = (user_agent or "*").lower()
    matched = [
        group
        for group in groups
        if any(agent == "*" or ua in agent or agent in ua for agent in group["agents"])
    ]
    return matched or [group for group in groups if "*" in group["agents"]]


def is_robots_allowed(text: str, url: str, user_agent: str = "*") -> bool:
    path = urlparse(url).path or "/"
    groups = parse_robots_txt(text, user_agent)
    if not groups:
        return True
    decision = True
    best = -1
    for group in groups:
        for rule in group["disallow"]:
            if rule and path.startswith(rule) and len(rule) >= best:
                decision = False
                best = len(rule)
        for rule in group["allow"]:
            if rule and path.startswith(rule) and len(rule) >= best:
                decision = True
                best = len(rule)
    return decision
